### PR TITLE
Cleanup & Fixes: TIO, Keys, Beats, TrackDAO

### PR DESCRIPTION
**Pull request by uklotzde**
Lots of minor cleanups, fixes and improvements around the mentioned classes. Some of the modifications depend on others so I combined them in a single PR.

### CODE REVIEW

File: src/track/beatgrid.cpp [hunk 1]
```diff
@@ -67,27 +67,26 @@ void BeatGrid::setGrid(double dBpm, double dFirstBeatSample) {
...
 
-QByteArray* BeatGrid::toByteArray() const {
+std::unique_ptr<QByteArray> BeatGrid::toByteArray() const {
     QMutexLocker locker(&m_mutex);
     std::string output;
     m_grid.SerializeToString(&output);
-    QByteArray* pByteArray = new QByteArray(output.data(), output.length());
-    // Caller is responsible for delete
-    return pByteArray;
+    return std::make_unique<QByteArray>(output.data(), output.length());
```
Comment (daschuer) [inline src/track/beatgrid.cpp]: I think we can improve this even more. 
This function is only used here 
https://github.com/mixxxdj/mixxx/blob/87ef849b94e6080b307fff33793567682c0687a2/src/library/dao/trackdao.cpp#L1605
To finally create a temorary QVariant. 

How about
```
QVariant BeatGrid::toVariant() const {
     QMutexLocker locker(&m_mutex);
     std::string output;
     m_grid.SerializeToString(&output);
     return QVariant(output.c_str);
}
```

File: src/track/beatgrid.cpp [hunk 2]
```diff
@@ -245,18 +244,18 @@ bool BeatGrid::findPrevNextBeats(double dSamples,
...
     }
     // qDebug() << "BeatGrid::findBeats startSample" << startSample << "stopSample"
     //          << stopSample << "beatlength" << m_dBeatLength << "BPM" << bpm();
     double curBeat = findNextBeat(startSample);
     if (curBeat == -1.0) {
-        return NULL;
+        return std::unique_ptr<BeatIterator>();
     }
-    return new BeatGridIterator(m_dBeatLength, curBeat, stopSample);
+    return std::make_unique<BeatGridIterator>(m_dBeatLength, curBeat, stopSample);
```
Comment (daschuer) [inline src/track/beatgrid.cpp]: It feels weird to delete an Iterator, but I understand that how it currently works. 

It is only used here
https://github.com/mixxxdj/mixxx/blob/49aaf97de17612b57d2a7618fcae0e687dbea8e0/src/waveform/renderers/waveformrenderbeat.cpp#L52
Cant we return the BeatGridIterator by value and rely on the Compilers return value optimization that avoids a deep copy?

Comment (uklotzde) [inline src/track/beatgrid.cpp]: BeatIterator is an abstract class and could not returned by value. The 
interface is defined by the base class.

On 02/11/2016 09:06 AM, Daniel Schürmann wrote:
In src/track/beatgrid.cpp 
<https://github.com/mixxxdj/mixxx/pull/892#discussion_r52571750>:

**Comment by daschuer** [comment]
Thank you for another nice clean up. LGTM

**Pull request merged**